### ISSUE TITLE
Flaky dtests in TestAlternatorStreams

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#15260]
In https://github.com/scylladb/scylla-dtest/issues/3498 we noticed that different dtests in `alternator_streams_test.py`::`TestAlternatorStreams` fail when trying to write to a table with Alternator Streams (i.e., CDC) enabled. We thought that the problem was that a user is not allowed to write to a table which has CDC before its first "generation" is ready. 

The supposed-to-be solution had two parts:
1. In Scylla PR https://github.com/scylladb/scylladb/pull/6901 @elcallio fixed Alternator to ensure `cdc::metadata::streams_available()` before reporting that Alternator Streams is enabled.
2. In dtest PR https://github.com/scylladb/scylla-dtest/pull/3513 I fixed the dtest to wait for the stream to become available before starting to write.

These two fixes should have solved the test failure, but apparently didn't - in the new run https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3426/ we see the failure https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3426/testReport/junit/alternator_stream_tests/TestAlternatorStreams/Tests___Sanity_Tests___test_added_node_gets_closed_shards/ :

```
self = <alternator_stream_tests.TestAlternatorStreams object at 0x7f0f74a47110>

    @pytest.mark.next_gating
    def test_added_node_gets_closed_shards(self):
        """
        test scenario:
            1. create a table with Streams.
            2. run alternator stress. (or multiple stresses to multiple nodes needed?)
            3. add new node to cluster.
            4. wait for new node bootstrap.
            5. run streams APIs queries, connecting to the new node.
            6. see that it returns some 'closed' shards with EndingSequenceNumber.
            https://github.com/scylladb/scylla/pull/8209#issuecomment-790625323
        """
        stream_specification = enums.StreamSpecification.KEYS_ONLY.value
        self.prepare_dynamodb_cluster(num_of_nodes=3)
        node1 = self.cluster.nodelist()[0]
    
        logger.info(f'Pre setup - Creating "{TABLE_NAME}" table via "{node1.name}" node with "{stream_specification}"'
                    f' stream key')
>       stream_arn = self.prefill_dynamodb_table(node=node1, stream_specification=stream_specification)[0]

alternator_stream_tests.py:253: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
alternator_utils.py:880: in prefill_dynamodb_table
    self.batch_write_actions(table_name=table_name, node=node, new_items=new_items)
alternator_utils.py:395: in batch_write_actions
    raise error
alternator_utils.py:388: in batch_write_actions
    batch.put_item(item)
/usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:103: in put_item
    self._add_request_and_process({'PutRequest': {'Item': Item}})
/usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:112: in _add_request_and_process
    self._flush_if_needed()
/usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:139: in _flush_if_needed
    self._flush()
/usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:144: in _flush
    response = self._client.batch_write_item(
/usr/local/lib/python3.11/site-packages/botocore/client.py:535: in _api_call
    return self._make_api_call(operation_name, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.DynamoDB object at 0x7f0f738f3590>
operation_name = 'BatchWriteItem'
api_params = {'RequestItems': {'user_table': [{'PutRequest': {'Item': {'hash_key': {...}, 'x': {...}}}}, {'PutRequest': {'Item': {'...Request': {'Item': {'hash_key': {...}, 'x': {...}}}}, {'PutRequest': {'Item': {'hash_key': {...}, 'x': {...}}}}, ...]}}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        service_name = self._service_model.service_name
        history_recorder.record(
            'API_CALL',
            {
                'service': service_name,
                'operation': operation_name,
                'params': api_params,
            },
        )
        if operation_model.deprecated:
            logger.debug(
                'Warning: %s.%s() is deprecated', service_name, operation_name
            )
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        api_params = self._emit_api_params(
            api_params=api_params,
            operation_model=operation_model,
            context=request_context,
        )
        endpoint_url, additional_headers = self._resolve_endpoint_ruleset(
            operation_model, api_params, request_context
        )
        request_dict = self._convert_to_request_dict(
            api_params=api_params,
            operation_model=operation_model,
            endpoint_url=endpoint_url,
            context=request_context,
            headers=additional_headers,
        )
        resolve_checksum_context(request_dict, operation_model, api_params)
    
        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            model=operation_model,
            params=request_dict,
            request_signer=self._request_signer,
            context=request_context,
        )
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            maybe_compress_request(
                self.meta.config, request_dict, operation_model
            )
            apply_request_checksum(request_dict)
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context
            )
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            http_response=http,
            parsed=parsed_response,
            model=operation_model,
            context=request_context,
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.InternalServerError: An error occurred (InternalServerError) when calling the BatchWriteItem operation (reached max retries: 0): Internal server error: exceptions::invalid_request_exception (cdc: attempted to get a stream from an earlier generation than the currently used one. With CDC you cannot send writes with timestamps too far into the past, because that would break consistency properties (write timestamp: 2023/09/03 09:41:54, current generation started at: 2023/09/03 09:41:54))

/usr/local/lib/python3.11/site-packages/botocore/client.py:980: InternalServerError
```

This suggests that even though the test waited for the stream to become available as done in https://github.com/scylladb/scylladb/pull/6901, using `cdc::metadata::streams_available()`  (cdc/metadata.cc), the function `cdc::metadata::get_stream()` in the same file still complained that `if (it->first > ts)` is true, whatever it means. 

At first I thought maybe we had an off-by-one error (maybe we wait for the generation but need to wait a second or millisecond longer), but looking again into `cdc::metadata::get_stream()` I think maybe the problem is different - maybe for some reason another generation appears after the first one, racing with our write? If a CDC write calls get_stream() and can race with creation of a new generation, it sounds like this bug can appear. @elcallio do you think this is possible?

By the way, the error message talks about "you cannot send writes with timestamps too far into the past", but the check `it->first > ts` doesn't really check anything about being "too far" - it can be just a millisecond before `it->first` and report this error.

NOTE: Although this PR is fairly ancient and not rebased on a recent master, it seems the test did run the latest dtest master, and both of the above fixes are included in this run.

**Comment by kbr-scylla** [comment]
> maybe for some reason another generation appears after the first one, racing with our write?

Should not happen unless you bootstrap a new node concurrently.

> It does indeed seem like a new CDC generation suddenly appeared between timestamping the alternator write and doing cdc processing. @kbr-scylla Should that check really be a simply "greater than", and should we really delete old generations that quickly? Should we not apply some sort of window for the incoming timestamp being lower than current gen ts as well?

I think this discussion should have all the answers:
https://github.com/scylladb/scylladb/issues/7251

> We concluded on a call that it's safest if the current restrictions stay. The client can either use server-timestamps or, if their request gets rejected due to belonging to an earlier generation, simply repeat it with a higher timestamp (note that with this error the client knows that the write was a definite failure).

The correct thing to do when this error appears is to retry the write.

---------

But this error should not appear in this test at this place, if I understand correctly.
We bootstrap the 3-node cluster first, which creates 3 CDC generations.
Then -- as I understand from the original post -- `prefill_dynamodb_table` is supposed to wait until we reach the timestamp of the last generation before doing any write. So the error should not appear.

I suspect that this is again fallout from disabling gossip waits in dtests. It could be that the "get last CDC generation timestamp" operation doesn't observe the last (i.e. third) generation, but some earlier one (e.g. second), and waits for it. Then we start doing the writes. And only then the last generation timestamp propagates across the cluster (which normally would have happened during bootstrap, but it doesn't due to disabled ring delays) and causes some writes to fail.

**Comment by elcallio** [comment]
> I think this discussion should have all the answers: #7251
> 
> > We concluded on a call that it's safest if the current restrictions stay. The client can either use server-timestamps or, if their request gets rejected due to belonging to an earlier generation, simply repeat it with a higher timestamp (note that with this error the client knows that the write was a definite failure).
> 
> The correct thing to do when this error appears is to retry the write.
> 

The problem here is that the time stamp _is_ server generated. In alternator when building the mutations. Which technically I guess makes it also client generated. In any case there is no guarantee this time stamp is not referencing an "old" cdc generation when we reach the `get_streams` call. 

> But this error should not appear in this test at this place, if I understand correctly. We bootstrap the 3-node cluster first, which creates 3 CDC generations. Then -- as I understand from the original post -- `prefill_dynamodb_table` is supposed to wait until we reach the timestamp of the last generation before doing any write. So the error should not appear.
> 
> I suspect that this is again fallout from disabling gossip waits in dtests. It could be that the "get last CDC generation timestamp" operation doesn't observe the last (i.e. third) generation, but some earlier one (e.g. second), and waits for it. Then we start doing the writes. And only then the last generation timestamp propagates across the cluster (which normally would have happened during bootstrap, but it doesn't due to disabled ring delays) and causes some writes to fail.

Well, the alternator stream "status" reported is based on "is there a valid cdc generation", and at some point here, before all nodes are up, there probably is one. Which gets superseded. Regardless of gossip-wait in tests, this sounds like a valid scenario when adding/removing a node as well.

Question is: should this be handled on alternator level or on CDC level? I mean, we could allow writing to older generations for X time. We could also handle the error in alternator mutate (assuming the error can be easily identified), or maybe alternator could stop using set timestamps? But I assume there is a very good reason for the latter.

**Comment by kbr-scylla** [comment]
> The problem here is that the time stamp is server generated. In alternator when building the mutations. Which technically I guess makes it also client generated. In any case there is no guarantee this time stamp is not referencing an "old" cdc generation when we reach the get_streams call.

Hmm, ok. It can happen if there's a yield between generating the timestamp and `get_streams`. Which is more likely the "further" we get from the low-level CDC code.

> Question is: should this be handled on alternator level or on CDC level? I mean, we could allow writing to older generations for X time. We could also handle the error in alternator mutate (assuming the error can be easily identified), or maybe alternator could stop using set timestamps? But I assume there is a very good reason for the latter.

The "we have lots of time and nothing better to do" way is indeed to modify the logic to allow writing to past generations for some time. But it needs to be given some thought and attention regarding the implications of this (won't this cause inconsistencies? and so on) before writing any code.

The lazy solution is to have the client retry the operation. Which is safe because this error is a definite failure -- if you get this particular error, you know that the write did not affect the cluster in any way. Maybe not the best UX but very simple thing to do.

**Comment by elcallio** [comment]
Problem with solution 2 is that we'll need to invent a reasonably unique exception (or error code?) to enable the client (alternator) to distinguish the case of a real failure. Assuming we can even propagate the base exception that far in all cases (?)... But probably still less impactful on behaviour than 1.). 

But regarding 1.), assuming we allow generations to be found for, say, 30s past being obsoleted, is not the worst thing that can happen that a write timestamped older could be missed if a client decides an old gen is no longer relevant (prematurely)?

**Comment by kbr-scylla** [comment]
> is not the worst thing that can happen that a write timestamped older could be missed if a client decides an old gen is no longer relevant (prematurely)?

Yes, missing writes (CDC log entries) is one problem I'm aware of.

That's why CDC drivers have this "confidence window" thing where they ignore most recent writes.

Not sure if crossing generation boundaries makes any difference, though.

**Comment by elcallio** [comment]
Well, typically, you'll iterate CDC generations and round-robin the ring for data until some time when you decide the generation no longer applies. So it sort of does, since if the cdc gen expiry + your own data time skew ends up larger than this gen discard leeway you'll miss the data. 
In the alternator streams case it would mean we get entries into the very sparsely populated generations of the cluster boot. But we should see it. 
I can try playing around with it, though I should probably see if I can even provoke the issue above locally first.

**Comment by elcallio** [comment]
Though I agree (just to make it clear) that since this is a rare condition, re-trying is first prio - we will need to do this anyway, since we could always be suspended in time and expire the use-old-gap.

**Comment by nyh** [comment]
> Hmm, ok. It can happen if there's a yield between generating the timestamp and `get_streams`. Which is more likely the "further" we get from the low-level CDC code.

I don't know when the get_streams happens, but note that Alternator writes in the test use LWT, so quite a bit of time may pass between choosing the timestamp by the LWT protocol and actually writing it (when I say "quite a bit of time", I mean milliseconds, not much more than that, but still yields should be surprising).

**Comment by elcallio** [comment]
The problem is that this does not handle the underlying problem. While not likely, we can have the exact same problem arise if a node is added/removed while doing writes to alternator with streams enabled, making a write cross the generation time stamp like this. Handling the error and retrying at the call site handles this as well, not only the test failure cause.

**Comment by nyh** [comment]
I just realized something: Because Scylla throws some CQL-specific error (invalid_request_exception) from the writing code, which Alternator doesn't understand, Alternator generates an InternalServerError in this case (as can indeed be seen above).  When a client receives an InternalServerError the client library is expected to retry the request. Because in Alternator the client doesn't supply the timestamp, the new request will get a new timestamp, and should succeed.

But the driver did not retry, and it said "reached max retries: 0". 
I see in dtest's alternator_utils.py:

```
GLOBAL_CONFIG = botocore.client.Config(retries={"max_attempts": 0}, read_timeout=300)
```

Maybe this max_attempts=0 is the whole problem!

**Comment by elcallio** [comment]
True, retrying on client level is perhaps even better.

**Comment by nyh** [comment]
Created a dtest PR (just a two-byte change ;-)) in https://github.com/scylladb/scylla-dtest/pull/3549 to return the number of retries in botocore (the AWS library used for Alternator tests) from 0 to the original default of 5.

If a write races with the topology change and fails on this InternalServerError, the driver will just silently retry it, up to 5 times, and it should fix this bug. At least I hope so - I don't know how to reliably reproduce this problem so I just checked my patch doesn't break the test...

**Comment by nyh** [comment]
Let's hope that the dtest fix (enabling retries) **really** fixes this issue. If we see more failures of these tests, we'll need to reopen yet again :-(

**Comment by fruch** [comment]
still happening:
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4272/testReport/junit/alternator_stream_tests/TestAlternatorStreams/FullDtest___full_split014___test_sequence_numbers_during_add_decommission_node/

@nyh I think  we should remove the speedup, for all streams cases if we can't figure it out.

**Comment by fruch** [comment]
@nyh 
we have test coverage and this is part of it:

`test_verify_stream_records_after_topology_changed`
`test_updated_shards_during_add_decommission_node`
`test_sequence_numbers_during_add_decommission_node`

if you think they aren't enough, please do specify what's missing, or what's wrong with those.

**Comment by fruch** [comment]
we have this in some of those tests:
```
self.run_write_stress(table_name=TABLE_NAME, node=node1, num_of_item=1000, ignore_errors=True)
```
which not necessary gonna work across all of the test, and doesn't fail the test at any error

**Comment by nyh** [comment]
> we have this in some of those tests:
> 
> ```
> self.run_write_stress(table_name=TABLE_NAME, node=node1, num_of_item=1000, ignore_errors=True)
> ```

Oh, I saw that but I didn't realize that this happens in a background thread :-( 

> which not necessary gonna work across all of the test, and doesn't fail the test at any error

Right. I don't know how much time writing 1000 items takes, and the "ignore_errors=True" means it can't recognize the problem of this issue - where writes fail.

So I guess a new test to try to reproduce this issue reliably is in order. It might not be easy at all (or might not even be reliably reproducible...)

**Comment by elcallio** [comment]
> > 
> 
> Note how 5 retries were actually done. @elcallio do you have any guess how we can have 5 consecutive failures of this? I'm not sure how frequently the boto driver retries the same query (I can try to take a look at their retry code...) but even if it's very frequent, I don't understand how more than one can fail - if the first one races with the topology change, wouldn't the second write automatically "win" the race because it starts after the topology change?

Assuming the alternator code creates a new time stamp on each retry, yes, I would agree...

**Comment by fruch** [comment]
happened on two different tests on latest next gating:

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4367/testReport/junit/alternator_stream_tests/TestAlternatorStreams/

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4367/testReport/junit/alternator_stream_tests/TestAlternatorStreams/FullDtest___full_split006___test_added_node_gets_closed_shards/

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4367/testReport/junit/alternator_stream_tests/TestAlternatorStreams/FullDtest___full_split013___test_updated_shards_during_add_decommission_node/

those test are going out of gating for now.

**Comment by fruch** [comment]
PR to take those tests out of gating:
https://github.com/scylladb/scylla-dtest/pull/3562

**Comment by elcallio** [comment]
What is the retry-backoff time for the driver? Could we be seeing that the alternator endpoint tries to do write ops that end up on a different coordinator node, and we have enough time skew between the machines that the timestamp generated on the alternator node is still in the past for the coordinator/cdc gen node?
@nyh Why do we use explicit timestamps in the generated mutations for batchwrite?

**Comment by elcallio** [comment]
For 3, I meant that it might be an idea if the mutations generated by alternator could have their server-side timestamp generated by the actual coordinator for the mutation execution, rather than the alternator endpoint (since they can be different). But at the very least here, and maybe in the future, this should not matter really.

**Comment by nyh** [comment]
> For 3, I meant that it might be an idea if the mutations generated by alternator could have their server-side timestamp generated by the actual coordinator for the mutation execution, rather than the alternator endpoint (since they can be different). But at the very least here, and maybe in the future, this should not matter really.

Note that in the LWT case (which this test uses), the timestamp **is** chosen by the regular LWT framework, and not by special code in Alternator. Alternator couldn't choose a timestamp in this case even if it wants to (just like you can't use USING TIMESTAMP in LWT writes in CQL). 

Moreover, when not using LWT, Alternator indeed creates the mutation itself with `api::new_timestamp()`. But in Alternator serves as a coordinator for this request and does the same timestamp-invention that a CQL coordinator would do - Alternator doesn't send the request to yet another "coordinator" which could have picked the timestamp (unless I'm misunderstanding what you meant).

**Comment by nuivall** [comment]
Can we enable cdc trace logs when running this test on CI? Perhaps it would give more data. One thing which interests me is how _gens changes, is it just that after adding a node one generation is added there?

If during retries _gens content for some reason changes that'd explain why client retries don't fix the problem.

**Comment by nyh** [comment]
> Can we enable cdc trace logs when running this test on CI? Perhaps it would give more data. One thing which interests me is how _gens changes, is it just that after adding a node one generation is added there?

We have several dtests that do something like
```
session = self.prepare(jvm_args=['--logger-log-level', 'prepared_statements_cache=trace', '--smp', '1'])
```

Before doing that you'll need to check that:
1. The trace-level messages that you hope exist, actually exist, and
2. the amount of spam generated by the trace log messages isn't excessive.

Another alternative is to recognize the interesting and relatively-rare events and log them at INFO level. I think this level is shown by default in dtest.

**Comment by fruch** [comment]
@nyh can you reopen this one ? I think it's a mistake that it was closed

**Comment by mykaul** [comment]
Is it being worked on?

**Comment by nyh** [comment]
I'll take this as a hint that you want me to work on it. I'll take a look.
CC @eliransin

**Comment by nyh** [comment]
@fruch can you please help me figure out how to know which dtests in alternator_streams_test.py::TestAlternatorStreams have been failing and when? Either in Jenkins or I saw in the past that you have some sort of elasticsearch setup to do this.
Do we have this test-searching mechanism documented somewhere?

**Comment by fruch** [comment]
> @fruch can you please help me figure out how to know which dtests in alternator_streams_test.py::TestAlternatorStreams have been failing and when? Either in Jenkins or I saw in the past that you have some sort of elasticsearch setup to do this. Do we have this test-searching mechanism documented somewhere?

it's not documented how to use it
let's talk next week, I'll show it, and let's try documenting it together (maybe let's record it)

**Comment by nyh** [comment]
@mykaul @fruch if this issue is to remain in P1, I think someone who knows how should make an effort to find relatively recent examples of failures. The last example I have is a link from more than a month ago, and the link no longer works. Thanks.

**Comment by fruch** [comment]
> @mykaul @fruch if this issue is to remain in P1, I think someone who knows how should make an effort to find relatively recent examples of failures. The last example I have is a link from more than a month ago, and the link no longer works. Thanks.

last spotted occurence is on Sep 10

https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/5f5a5870-6b31-11ee-bf28-5927cc3d3ce3

but keep in mind, it was taken out of gating, so we have much less runs of it.

**Comment by nyh** [comment]
> > @mykaul @fruch if this issue is to remain in P1, I think someone who knows how should make an effort to find relatively recent examples of failures. The last example I have is a link from more than a month ago, and the link no longer works. Thanks.
> 
> last spotted occurence is on Sep 10
> 
> https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/5f5a5870-6b31-11ee-bf28-5927cc3d3ce3

Thanks, this is useful for finding the names of the failing tests, I'll have to see if any of them has any details saved (one link I followed from above, didn't).
Why is there nothing before August 11, do we not have records before then, or it really never failed before then?

@avikivity and @eliransin said we were running all dtests on modern master 100 times. Were these tests also running there?  Did they fail?

**Comment by fruch** [comment]
> > > @mykaul @fruch if this issue is to remain in P1, I think someone who knows how should make an effort to find relatively recent examples of failures. The last example I have is a link from more than a month ago, and the link no longer works. Thanks.
> > 
> > last spotted occurence is on Sep 10
> > 
> > https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/5f5a5870-6b31-11ee-bf28-5927cc3d3ce3
> 
> Thanks, this is useful for finding the names of the failing tests, I'll have to see if any of them has any details saved (one link I followed from above, didn't).
> Why is there nothing before August 11, do we not have records before then, or it really never failed before then?
> 

I'm looking for specific failures (see the query)

> @avikivity and @eliransin said we were running all dtests on modern master 100 times. Were these tests also running there?  Did they fail?
> 

We are running gating a lot, but what not in gating, only runs once a day at most (some days the build fails and there no daily runs)

So we took those out of the way, which also get us from having it reproduced and figured out.

We can create a job and run this specific (all of alternator streams) suite again and again to try to flush it out.

**Comment by nyh** [comment]
> > > > @mykaul @fruch if this issue is to remain in P1, I think someone who knows how should make an effort to find relatively recent examples of failures. The last example I have is a link from more than a month ago, and the link no longer works. Thanks.
> > > 
> > > 
> > > last spotted occurence is on Sep 10
> > > https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/5f5a5870-6b31-11ee-bf28-5927cc3d3ce3
> > 
> > 
> > Thanks, this is useful for finding the names of the failing tests, I'll have to see if any of them has any details saved (one link I followed from above, didn't).
> > Why is there nothing before August 11, do we not have records before then, or it really never failed before then?
> 
> I'm looking for specific failures (see the query)

Yes, but I was asking if it's some sort of regression - never failed before August 11th, or simply the recording of the data started in August.

> > @avikivity and @eliransin said we were running all dtests on modern master 100 times. Were these tests also running there?  Did they fail?
> 
> We are running gating a lot, but what not in gating, only runs once a day at most (some days the build fails and there no daily runs)

I was under the impression that somebody from QA was running these days all dtests 100 times to see the results. I didn't realize this was limited to "gating" tests. Is it limited only to gating?

I don't understand why this test is in P1 if we already decided to move it out of gating and it hasn't failed for a month. Yes, it would be great to figure it out eventually, but P1, whose description is "urgent"?

**Comment by fruch** [comment]
> > > > > @mykaul @fruch if this issue is to remain in P1, I think someone who knows how should make an effort to find relatively recent examples of failures. The last example I have is a link from more than a month ago, and the link no longer works. Thanks.
> > > > 
> > > > 
> > > > last spotted occurence is on Sep 10
> > > > https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/5f5a5870-6b31-11ee-bf28-5927cc3d3ce3
> > > 
> > > 
> > > Thanks, this is useful for finding the names of the failing tests, I'll have to see if any of them has any details saved (one link I followed from above, didn't).
> > > Why is there nothing before August 11, do we not have records before then, or it really never failed before then?
> > 
> > 
> > I'm looking for specific failures (see the query)
> 
> Yes, but I was asking if it's some sort of regression - never failed before August 11th, or simply the recording of the data started in August.
> 
> > > @avikivity and @eliransin said we were running all dtests on modern master 100 times. Were these tests also running there?  Did they fail?
> > 
> > 
> > We are running gating a lot, but what not in gating, only runs once a day at most (some days the build fails and there no daily runs)
> 
> I was under the impression that somebody from QA was running these days all dtests 100 times to see the results. I didn't realize this was limited to "gating" tests. Is it limited only to gating?

No one from QA is rerunning tests manually/automatic like that, if one needs it's from investigation of on of those
a. he can do it on his own
b. if doesn't know, he should be asking for help with it.

nothing is happening "automatically" around those.

> 
> I don't understand why this test is in P1 if we already decided to move it out of gating and it hasn't failed for a month. Yes, it would be great to figure it out eventually, but P1, whose description is "urgent"?

**Comment by nyh** [comment]
> > I was under the impression that somebody from QA was running these days all dtests 100 times to see the results. I didn't realize this was limited to "gating" tests. Is it limited only to gating?
> 
> No one from QA is rerunning tests manually/automatic like that, if one needs it's from investigation of on of those a. he can do it on his own b. if doesn't know, he should be asking for help with it.

I understood from @avikivity and @eliransin that last week there was a coordinated effort to run **all** tests (not specific tests that someone needs to ask for) 100 times with the current master. Did I misunderstand?

I know how to run a single dtest (in this case, all the tests of a specific class). I also know how to run it in a loop on my laptop. I just don't how to automate this to Jenkins, and it's not my place to tell you to do it. I'm just exploring all my options.

**Comment by mykaul** [comment]
@nyh - @yaronkaikov is now analyzing the results of that effort.

**Comment by bhalevy** [comment]
@nyh 
The test seems stable now in release mode but still flaky in debug mode.
See https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-debug/274/testReport/alternator_stream_tests/TestAlternatorStreams/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split070___test_sequence_numbers_during_add_decommission_node/
```
AssertionError: No Leaving node found, decommission is not running
assert 'UL' in "Datacenter: datacenter1\n=======================\nStatus=Up/Down\n|/ State=Normal/Leaving/Joining/Moving\n--  Address     Load       Tokens       Owns    Host ID                               Rack\nUN  127.0.58.1  23.28 MB   256          ?       b0e94bb3-7416-40e9-ac96-84809b714ac2  rack1\nUN  127.0.58.2  24.4 MB    256          ?       d58b2153-4407-4f6a-88eb-3e93bbe6568b  rack1\nUN  127.0.58.3  23.96 MB   256          ?       84b9d8cc-63e4-4e3b-a656-199ca03fa5cf  rack1\nUJ  127.0.58.6  ?          256          ?       null                                  rack1\n\nNote: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless\n"
```

**Comment by bhalevy** [comment]
> @nyh The test seems stable now in release mode but still flaky in debug mode. See https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-debug/274/testReport/alternator_stream_tests/TestAlternatorStreams/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split070___test_sequence_numbers_during_add_decommission_node/
> 
> ```
> AssertionError: No Leaving node found, decommission is not running
> assert 'UL' in "Datacenter: datacenter1\n=======================\nStatus=Up/Down\n|/ State=Normal/Leaving/Joining/Moving\n--  Address     Load       Tokens       Owns    Host ID                               Rack\nUN  127.0.58.1  23.28 MB   256          ?       b0e94bb3-7416-40e9-ac96-84809b714ac2  rack1\nUN  127.0.58.2  24.4 MB    256          ?       d58b2153-4407-4f6a-88eb-3e93bbe6568b  rack1\nUN  127.0.58.3  23.96 MB   256          ?       84b9d8cc-63e4-4e3b-a656-199ca03fa5cf  rack1\nUJ  127.0.58.6  ?          256          ?       null                                  rack1\n\nNote: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless\n"
> ```

That just because it takes so much longer to bootstrap a new node (with vnodes) in debug mode

**Comment by nyh** [comment]
> @mykaul
> 
> One of the pushes you did to your fork, has closed this issue
> 
> You probably want to reopen it

@avikivity this is the danger of references like "Fixes scylladb/scylla#15260". See https://github.com/mykaul/scylla-dtest/commit/2f106d4e9071af25e65780748cacb14523ee65ac

**Comment by mykaul** [comment]
> > @mykaul
> > One of the pushes you did to your fork, has closed this issue
> > You probably want to reopen it
> 
> @avikivity this is the danger of references like `Fixes scylladb/scylla#15260`. See [mykaul/scylla-dtest@2f106d4](https://github.com/mykaul/scylla-dtest/commit/2f106d4e9071af25e65780748cacb14523ee65ac)

ARGH! How did that happen?!

**Comment by fruch** [comment]
> > > @mykaul
> > > One of the pushes you did to your fork, has closed this issue
> > > You probably want to reopen it
> > 
> > @avikivity this is the danger of references like `Fixes scylladb/scylla#15260`. See [mykaul/scylla-dtest@2f106d4](https://github.com/mykaul/scylla-dtest/commit/2f106d4e9071af25e65780748cacb14523ee65ac)
> 
> ARGH! How did that happen?!

Never work on master/next in your fork

**Comment by fruch** [comment]
> > > > @mykaul
> > > > One of the pushes you did to your fork, has closed this issue
> > > > You probably want to reopen it
> > > 
> > > @avikivity this is the danger of references like `Fixes scylladb/scylla#15260`. See [mykaul/scylla-dtest@2f106d4](https://github.com/mykaul/scylla-dtest/commit/2f106d4e9071af25e65780748cacb14523ee65ac)
> > 
> > ARGH! How did that happen?!
> 
> Never work on master/next in your fork 

There are more reasons for that, this is one example, github doesn't tell the difference between the main project and it's forks when it comes to closing issues

**Comment by nyh** [comment]
@fruch should we reopen this issue because it was accidentally closed? Have we seen more failures in TestAlternatorStreams since September (I don't know how to search for it)?

**Comment by fruch** [comment]
@nyh 

see
https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/39e754a0-b053-11ee-81c7-3986d18dafd5

those tests didn't failed in the last 30 day on master (one of them did failed once on 5.2/2022.2)

**Comment by bhalevy** [comment]
So why reopen this issue?

**Comment by fruch** [comment]
the issue was accidentally closed, after it was reopened. that's why it's reopen.

someone would still need to do the work of check those specific test to be stable, and take a decision, until them issue is open.

**Comment by fruch** [comment]
@nyh 

there are still quite a bunch of tests in dtest naming this issue, as reason for remove test from gating

I would recommend to put them be into gating
